packages/flutter/lib/src/material/animated_raster.dart

```diff
@@ -0,0 +1,202 @@
+  /// Create a new [_RenderAnimatedRaster].
+  _RenderAnimatedRaster(
+    this._animation,
+    this._delegate,
+    this._devicePixelRatio,
+  );
+
+  /// The delegate that controls how the child raster is painted.
+  _AnimatedRasterDelegate get delegate => _delegate;
+  _AnimatedRasterDelegate _delegate;
+  set delegate(_AnimatedRasterDelegate value) {
+    if (value == delegate) {
+      return;
+    }
+    _delegate = value;
+    markNeedsPaint();
+  }
+
+  /// The device pixel ratio used to create the child raster.
+  double get devicePixelRatio => _devicePixelRatio;
+  double _devicePixelRatio;
+  set devicePixelRatio(double value) {
+    if (value == devicePixelRatio) {
+      return;
+    }
+    _devicePixelRatio = value;
+    markNeedsPaint();
+  }
+
+  /// The animation used to drive this render object.
+  Animation<double> get animation => _animation;
+  Animation<double> _animation;
+  set animation(Animation<double> value) {
+    if (value == animation) {
+      return;
+    }
+    animation.removeListener(markNeedsPaint);
+    _animation = value;
+    animation.addListener(markNeedsPaint);
+    markNeedsPaint();
+  }
+
+  @override
+  void dispose() {
+    animation.removeListener(markNeedsPaint);
+    _childImage?.dispose();
+    _childImage = null;
+    super.dispose();
+  }
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    animation.addListener(markNeedsPaint);
+    super.attach(owner);
+  }
+
+  @override
+  void detach() {
+    animation.removeListener(markNeedsPaint);
+    super.detach();
+  }
+
+  // Paint [child] with this painting context, then convert to a raster and detach all
+  // children from this layer.
+  ui.Image _paintAndDetachToGpuImage() {
+    final OffsetLayer offsetLayer = layer! as OffsetLayer;
+    final PaintingContext context = PaintingContext(offsetLayer, Offset.zero & size);
+    context.paintChild(child!, Offset.zero);
+    // ignore: invalid_use_of_protected_member
+    context.stopRecordingIfNeeded();
+    final ui.Image image = offsetLayer.toGpuImage(Offset.zero & size, pixelRatio: devicePixelRatio);
+    offsetLayer.removeAllChildren();
+    return image;
+  }
+
+  @override
+  bool get isRepaintBoundary => child != null;
+
+  ui.Image? _childImage;
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child == null) {
+      return;
+    }
+    final bool useRaster = delegate.useRaster(animation);
+    if (!useRaster) {
+      delegate.paint(
+        context,
+        animation,
+        offset & size,
+        super.paint,
+      );
+      _childImage?.dispose();
+      _childImage = null;
+      return;
+    }
+    _childImage ??= _paintAndDetachToGpuImage();
+    try {
+      delegate.paintRaster(
+        context,
+        _childImage!,
+        devicePixelRatio,
+        animation,
+      );
+    } on ui.PictureRasterizationException {
+      // We're running on an emulator without a GPU context.
+      // For now render nothing.
```

Comment:
> So far this only happens on android emulators with --use-software-rendering. 

Seems also happens when using `flutter test` for widget tests, i.e. without a simulator/real-device?

Reply:
Flutter test won't throw here, but it currently will return a checkerboard image.

We could add API that tells you whether this API is available or not to avoid this -my assumption with the current design is that developers would be in a place here to just draw their original picture or scene instead, but maybe that's not a good assumption.